### PR TITLE
esp32/boards/TINYPICO: Added helper libraries

### PR DESCRIPTION
**Pull request by UnexpectedMaker**
Added a new modules folder in the TinyPICO board folder that has the extra TinyPICO specific modules I'd like frozen with the build.

### CODE REVIEW

File: ports/esp32/boards/TINYPICO/modules/tinypico.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+# Helper functions
+
+# Get a *rough* estimate of the current battery voltage
+# If the battery is not present, the charge IC will still report it's trying to charge at X voltage
+# so it will still show a voltage.
+def get_battery_voltage():
+    """
+    Returns the current battery voltage. If no battery is connected, returns 3.7V
+    This is an approximation only, but useful to detect of the chnare state of the battery is getting low.
```
Comment (nevercast) [inline ports/esp32/boards/TINYPICO/modules/tinypico.py]: s/chnare/charge

File: ports/esp32/boards/TINYPICO/modules/tinypico.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+def get_internal_temp_F():
+    """Return the internal PICO-D4 temperature in farenheit."""
+    return esp32.raw_temperature()
+
+# Return the internal PICO-D4 temperature in Celcius
+def get_internal_temp_C():
+    """Return the internal PICO-D4 temperature in celcius."""
+    return ( esp32.raw_temperature() - 32 ) / 1.8
+
+# Power to the on-oard Dotstar is controlled by a PNP transistor, so low is ON and high is OFF
```
Comment (nevercast) [inline ports/esp32/boards/TINYPICO/modules/tinypico.py]: s/on-oard/on-board

File: ports/esp32/boards/TINYPICO/modules/tinypico.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        Pin( DOTSTAR_PWR, Pin.OUT, None )   # Break the PULL_HOLD on the pin
+        Pin( DOTSTAR_PWR ).value( False )   # Set the pin to LOW to enable the Transistor
+    else:
+        Pin(13, Pin.IN, Pin.PULL_HOLD) # Set PULL_HOLD on the pin to allow the 3V3 pullup to work
+                   
+    Pin(DOTSTAR_CLK, Pin.OUT if state else Pin.IN )     # If power is on, set CLK to be output, otherwise input
+    Pin(DOTSTAR_DATA, Pin.OUT if state else Pin.IN )    # If power is on, set DATA to be output, otherwise input
+
+    # A small delay to let the IO change state
+    time.sleep(.035)
```
Comment (nevercast) [inline ports/esp32/boards/TINYPICO/modules/tinypico.py]: Did you observe a need for this in testing? I'd have thought the GPIO would change start faster than Python could return to the next opcode. Interesting!

File: ports/esp32/boards/TINYPICO/modules/tinypico.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    elif wheel_pos < 170:
+        wheel_pos -= 85
+        return (0, wheel_pos * 3, 255 - wheel_pos * 3)
+    else:
+        wheel_pos -= 170
+        return ( wheel_pos * 3, 255 - wheel_pos * 3, 0)
+
+# Go into deep sleep but shut down the APA first to save power
+# Use this if you want lowest deep  sleep current
+def go_deepsleep( t ):
```
Comment (nevercast) [inline ports/esp32/boards/TINYPICO/modules/tinypico.py]: Consider just `deepsleep`, avoid verbs like `do`, `calculate` etc when the noun makes the action obvious.

File: ports/esp32/boards/TINYPICO/modules/tinypico.py (hunk 3 above)
Comment (UnexpectedMaker) [reply ports/esp32/boards/TINYPICO/modules/tinypico.py]: Yes, it's needed unfortunately. It's ugly :(

File: ports/esp32/boards/TINYPICO/modules/tinypico.py (hunk 4 above)
Comment (UnexpectedMaker) [reply ports/esp32/boards/TINYPICO/modules/tinypico.py]: Sure, I will keep that in mind for the future, thanks!
This needs to stay compatible with the 2000+ TinyPICOs out there with the non-frozen helper so will leave it for now.

Comment (nevercast) [reply ports/esp32/boards/TINYPICO/modules/tinypico.py]: Deprecated? :yum: 
Makes sense to leave it for compatibility.

File: ports/esp32/boards/TINYPICO/modules/tinypico.py (hunk 3 above)
Comment (nevercast) [reply ports/esp32/boards/TINYPICO/modules/tinypico.py]: Ew :( my condolences.

**Comment by nevercast** [comment]
Flashed this `BOARD` to my TinyPico and can verify the modules work as expected.

**Review by nevercast — APPROVED** [review]
Builds, Flashes, Damien's 80MHz fix is working, `dotstar` and `tinypico` is present, Example code shipped with TinyPICO works as expected.

LGTM. :rocket: :heavy_check_mark:

**Review by nevercast — CHANGES_REQUESTED** [review]
The code that shipped with the TinyPICO messed with my test. Sorry, `dotstar` and `tinypico` are **NOT** present as expected.

@UnexpectedMaker Are these files present in your tests?

**Comment by nevercast** [comment]
Spoke with @UnexpectedMaker in Slack. @jimmo has plans to implement a solution for picking up these modules.

My thought would be replacing `FROZEN_MPY_DIR` with `FROZEN_MPY_DIRS` in `ports/esp32/Makefile`, and defaulting `FROZEN_MPY_DIRS` as `$FROZEN_MPY_DIR` in `py/mkrules.mk`.

Then in `py/mkrules.mk`, consume the list of dirs in to individual files.

In `boards/BOARD/mpconfigboard.mk`, one would add
```
FROZEN_MPY_DIRS += $(BOARD_DIR)/file.py
```

Unless you wanted to keep it more static and simple, @jimmo ?

**Comment by jimmo** [comment]
Yes, something like that, but I hadn't got around to thinking about it any more yet and need to take a closer look at what the other ports do. Possibly also need to consider what the resulting directory structure should look like. i.e. what happens if you have `foo.py` in multiple input directories.

**Comment by nevercast** [comment]
> what happens if you have `foo.py` in multiple input directories.

What I have at the moment is a script that runs before hand, the general rule is to trample in order. A directory defined later in the list tramples files from a previous one. This allows overriding files that ship with MicroPython for example, or that ship with that board.

**Comment by dpgeorge** [comment]
I merged the 80MHz patch in e9af6f5f886337d0a366ac4d7f4a03ee1117cde0

See #5084 for a proposal to have more sophisticated control over frozen scripts, which would be a prerequisite for the PR here.

File: ports/esp32/boards/TINYPICO/modules/tinypico.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        measuredVal += io.value()
+
+    return ( measuredVal == 0 )             # return True if the value is 0
+
+# Return the internal PICO-D4 temperature in Farenheit
+def get_internal_temp_F():
+    """Return the internal PICO-D4 temperature in farenheit."""
+    return esp32.raw_temperature()
+
+# Return the internal PICO-D4 temperature in Celcius
```
Comment (mcauser) [inline ports/esp32/boards/TINYPICO/modules/tinypico.py]: s/Celcius/Celsius

File: ports/esp32/boards/TINYPICO/modules/tinypico.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    """
+    measuredVal = 0                         # start our reading at 0
+    io = Pin( BAT_CHARGE, Pin.IN )          # Assign the pin to read
+  
+    for y in range(0, 10):                  # loop through 10 times adding the read values together to ensure no false positives    
+        measuredVal += io.value()
+
+    return ( measuredVal == 0 )             # return True if the value is 0
+
+# Return the internal PICO-D4 temperature in Farenheit
```
Comment (mcauser) [inline ports/esp32/boards/TINYPICO/modules/tinypico.py]: s/Farenheit/Fahrenheit

File: ports/esp32/boards/TINYPICO/modules/tinypico.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+def get_internal_temp_C():
+    """Return the internal PICO-D4 temperature in celcius."""
+    return ( esp32.raw_temperature() - 32 ) / 1.8
+
+# Power to the on-board Dotstar is controlled by a PNP transistor, so low is ON and high is OFF
+# We also need to set the Dotstar clock and data pins to be inputs to prevent power leakage when power is off
+# This might be improved at a future date
+# The reason we have power control for the Dotstar is that it has a quiescent current of around 1mA, so we
+# need to be able to cut power to it to minimise power consumption during deep sleep or with general battery powered use
+# to minimse un-needed battery drain
```
Comment (mcauser) [inline ports/esp32/boards/TINYPICO/modules/tinypico.py]: s/minimse/minimise

File: ports/esp32/boards/TINYPICO/modules/tinypico.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+__version__ = "0.0.0-auto.0"
+__repo__ = "https://github.com/TinyPICO/tinypico-micropython"
+
+# Import required libraries 
+from micropython import const
+from machine import Pin, SPI, ADC
+import machine, time, esp32
+
+# TinyPICO Hardare Pin Assingments
```
Comment (mcauser) [inline ports/esp32/boards/TINYPICO/modules/tinypico.py]: s/Hardare/Hardware
s/Assingments/Assignments

Comment (nevercast) [reply ports/esp32/boards/TINYPICO/modules/tinypico.py]: Wow I missed a lot of these.

File: ports/esp32/boards/TINYPICO/modules/tinypico.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    return ( esp32.raw_temperature() - 32 ) / 1.8
+
+# Power to the on-board Dotstar is controlled by a PNP transistor, so low is ON and high is OFF
+# We also need to set the Dotstar clock and data pins to be inputs to prevent power leakage when power is off
+# This might be improved at a future date
+# The reason we have power control for the Dotstar is that it has a quiescent current of around 1mA, so we
+# need to be able to cut power to it to minimise power consumption during deep sleep or with general battery powered use
+# to minimse un-needed battery drain
+def set_dotstar_power( state ):
+    """Set the power for the on-board Dostar to allow no current draw when not needed."""
```
Comment (nevercast) [inline ports/esp32/boards/TINYPICO/modules/tinypico.py]: s/Dostar/Dotstar

File: ports/esp32/boards/TINYPICO/modules/tinypico.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        return (255 - wheel_pos * 3, 0, wheel_pos * 3)   
+    elif wheel_pos < 170:
+        wheel_pos -= 85
+        return (0, wheel_pos * 3, 255 - wheel_pos * 3)
+    else:
+        wheel_pos -= 170
+        return ( wheel_pos * 3, 255 - wheel_pos * 3, 0)
+
+# Go into deep sleep but shut down the APA first to save power
+# Use this if you want lowest deep  sleep current
```
Comment (nevercast) [inline ports/esp32/boards/TINYPICO/modules/tinypico.py]: s/deep  sleep/deepsleep

Comment (nevercast) [inline ports/esp32/boards/TINYPICO/modules/tinypico.py]: Double space.

File: ports/esp32/boards/TINYPICO/modules/tinypico.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    This is an approximation only, but useful to detect of the charge state of the battery is getting low.
+    """
+    adc = ADC( Pin( BAT_VOLTAGE ) )         # Assign the ADC pin to read
+    measuredvbat = adc.read()               # Read the value
+    measuredvbat /= 4095                    # divide by 4095 as we are using the default ADC voltage range of 0-1V
+    measuredvbat *= 3.7                     # Multiply by 3.7V, our reference voltage
+    return measuredvbat
+
+# Return the current charge state of the battery - we need to read the value multiple times
+# to eliminate false negatives due to the charge IC not knowing the difference between no battery
```
Comment (nevercast) [inline ports/esp32/boards/TINYPICO/modules/tinypico.py]: Clarify with line #77

False negatives of False positives.

**Review by nevercast — COMMENTED** [review]
Just a few more typos and a clarification on wording.

**Comment by UnexpectedMaker** [comment]
Ok, this is all just plain embarrasing... going to hid under a rock somewhere :(

**Comment by nevercast** [comment]
s/hid/hide :)

**Comment by dpgeorge** [comment]
@UnexpectedMaker are you able to update this to use the new `FROZEN_MANIFEST` mechanism?

**Comment by nevercast** [comment]
Update blocked by https://github.com/micropython/micropython/pull/5218 ?

**Comment by dpgeorge** [comment]
> Update blocked by #5218 ?

That PR is now merged, so this one should be ready to update and merge as well.

**Comment by dpgeorge** [comment]
Superseded by #5257

**Pull request closed without merging**